From 22c15ccc6d1538455ae9e5cc22c0cc05a68d8c72 Mon Sep 17 00:00:00 2001
From: Jianing Ren <jianing.ren@rock-chips.com>
Date: Fri, 17 Apr 2020 18:26:03 +0800
Subject: [PATCH] usb: dwc3: core: fix drd bug for micro USB interface

Vbus detect function for INNO USB2PHY will be disabled when otg
cable is plugged. If we do not enable vbus detect when otg cable
unplugged, bvalid irq will never be triggered. Therefore, we set
phy mode to device when otg port is disconnected.

Change-Id: I5f53024d7a2a3a3f4f92ebcd716a25266fd009a1
Signed-off-by: Jianing Ren <jianing.ren@rock-chips.com>
---
 drivers/usb/dwc3/core.c | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/drivers/usb/dwc3/core.c b/drivers/usb/dwc3/core.c
index fecf801f7630..a51717be794a 100644
--- a/drivers/usb/dwc3/core.c
+++ b/drivers/usb/dwc3/core.c
@@ -264,6 +264,14 @@ static void __dwc3_set_mode(struct work_struct *work)
 disconnect:
 		switch (dwc->current_dr_role) {
 		case DWC3_GCTL_PRTCAP_HOST:
+			/*
+			 * We set device mode to disable otg-vbus supply and
+			 * enable vbus detect for inno USB2PHY.
+			 */
+			phy_set_mode(dwc->usb2_generic_phy,
+				     PHY_MODE_USB_DEVICE);
+			phy_set_mode(dwc->usb3_generic_phy,
+				     PHY_MODE_USB_DEVICE);
 			phy_power_off(dwc->usb3_generic_phy);
 			dwc3_host_exit(dwc);
 			break;
-- 
2.35.3

